woocommerce/woocommerce-gutenberg-products-block
# Fix flaw with event emitters and complete checkout flow for the cheque payment method.

With master having the necessary pieces in place to theoretically complete a payment using the cheque payment method and fulfill the checkout flow (including a redirect to order receipt page), I took some time to test things out and fix bugs that popped up.

In the process I discovered a _serious_ flaw with the event emitters that I overlooked when designing the system. The flaw is that all the registered observers on the event emitters _cannot react to any state changes_ because they are all invoked in one render process.  This also means that any state changes in observers can be problematic because subsequent observers would not know of those state changes.

The solution involves recognizing that we are using the event emitters for tracking flow through the checkout and to limit _status_ and _state_ changes to after the event has emitted.  By adding event emitters to the PaymentMethodData context, and adding another Checkout status it actually ended up simplifying the flow logic because now there is a stable point for payment methods to hook in their behaviour and for the `CheckoutProcessor` component to hook in it's logic. 

> Sidenote: I realize "simplifying" seems to be an oxymoron here because it's still relatively complex. But the complexity is absorbed by the providers instead of passed on to extension components to manage (which can increase complexity exponentially if extensions have to check constantly for status changes among all the elements in the checkout flow).

With the changes in this pull:

- Payment method extensions will not be responsible for updating the payment status. The payment method provider context handles that. Payment method components will register observers on the events they need to react to and return appropriate values for communicating with checkout.
- Convention (not enforced) will be that observers _must_ not set any state in checkout (they can set their own local state if they want). Provider state (payment, checkout, shipping) is handled by provider's emitters after all observers for an event have fired.

Here's a _rough_ diagram of the flow (not including shipping events):

![CheckoutFlow@2x](https://user-images.githubusercontent.com/1429108/78320562-bba49d00-7537-11ea-8487-c476454f1938.png)

In this pull:

- new payment method event emitters are added,
- new actions/status to checkout state context added (checkoutProcessingComplete)
- refactor checkout state provider.
- refactor checkout processor.
- wire up cheque payment method so it's hooked into the `onPaymentProcessing` event emitter.

## To Test

It is expected that Cart/Checkout behaviour up to the point of clicking the related submit button should continue to work as expected outside what was done in this branch.

You should be able to finish a complete checkout flow using the cheque payment method without errors.

There should be no errors in the console while testing.

Stripe is _not_ expected to work.

Validation errors after submit button is clicked _may_ have regressed. Please do test, but we can fix in follow-ups unless the cause is clear.


## Followups:

- since payment status is now controlled by checkout, I'm going to remove the payment status setter from being exposed on the `usePaymentMethodInterface` hook. That forces payment methods to use the appropriate api for hooking into the checkout flow (events).
- stripe client side will need modified to work with the new payment event emitters.
- I _may_ have broken some validation behaviour. If so, let's fix in follow-ups to avoid this pull becoming stale.
